funfact/algorithm.py

```diff
@@ -117,7 +119,15 @@ class _Factorization(Factorization, ab.AutoGradMixin):
             'Invalid optimization algorithm:\n{e}'
         )
 
-    loss_and_grad = ab.loss_and_grad(loss, opt_fac, target,
+    def loss_and_penalty(model, target, sum_vec=True, **kwargs):
+        if penalty_weight > 0:
```

Comment:
Based on the DRY (don't repeat yourself) principle, it is probably better to first compute the loss, and then conditionally compute the penalty term. This way, we don't repeat the call to the loss function.